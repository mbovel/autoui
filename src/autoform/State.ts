import { Tree, ensure } from "./utils";
import { Action } from "./Action";
import { set, toPath, isObject, omit, remove, unset, get, isArray } from "lodash-es";
import { JSONType } from "./utils";

export interface ActivePath {
	path: string;
	user: string;
}

export interface State<D extends JSONType = JSONType> {
	data: D;

	// Meta
	touched?: Tree<true | undefined>;
	active?: ActivePath[];
}

export function mutateState(state: State, action: Action): void {
	switch (action.type) {
		case "set":
			set(state, "data." + action.path, action.data);
			return;
		case "unset":
			unset(state, "data." + action.path);
			return;
		case "insertAfter":
			getArray(state, "data." + action.path).splice(action.index, 0, action.data);
			return;
		case "removeAt":
			getArray(state, "data." + action.path).splice(action.index, 1);
			return;
		case "sort":
			getArray(state, "data." + action.path).sort(action.compare);
			return;
		case "focus":
			state.active = state.active ?? [];
			state.active.push(omit(action, "type"));
			return;
		case "blur":
			state.active = state.active ?? [];
			remove(state.active, el => el.path !== action.path && el.user !== action.user);

			// Touch if not already touched
			if (isTouched(state.touched, action.path)) return;
			set(state, "touched." + action.path, true);

			return;
	}
}

export function isTouched(tree: Tree<true | undefined>, path: string): boolean {
	return !!toPath(path).reduce(getTouchedTreeChild, tree);
}

export function getTouchedTreeChild(node: Tree<true | undefined>, key: string | number) {
	return isObject(node) ? node[key] : node;
}

function getArray(object: any, path: string): Array<any> {
	const array = get(object, path);
	ensure(isArray(array));
	return array;
}
