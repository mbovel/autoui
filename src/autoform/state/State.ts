import { Primitive, isPrimitive, MapFunction, ensure } from "../utils";

import {
	memoize,
	mapValues,
	isArray,
	zipObject,
	isObjectLike,
	toPath,
	initial,
	last
} from "lodash-es";

import { JSON } from "../utils";

export interface BranchState {
	children: { [key: string]: State };
}

export interface MapState extends BranchState {
	type: "map";
}

export interface ListState extends BranchState {
	type: "list";
	order: string[];
}

export interface PrimitiveState<V> {
	touched: boolean;
	focusedBy: string[];
	value: V;
}

export interface StringState extends PrimitiveState<string> {
	type: "string";
}

export interface NumberState extends PrimitiveState<number> {
	type: "number";
}

export interface BooleanState extends PrimitiveState<boolean> {
	type: "boolean";
}

export type State = MapState | ListState | StringState | NumberState | BooleanState;

export function isBranchState(state: State): state is ListState | MapState {
	return state.type === "list" || state.type === "map";
}

export namespace State {
	export function fromJs(data: any, getKey: MapFunction<string>): State {
		if (isPrimitive(data)) {
			return { type: typeof data as any, value: data, touched: false, focusedBy: [] };
		} else if (isArray(data)) {
			const keys = data.map(getKey);
			const values = data.map(v => fromJs(v, getKey));
			return { type: "list", order: keys, children: zipObject(keys, values) };
		} else if (isObjectLike(data)) {
			return { type: "map", children: mapValues(data, v => fromJs(v, getKey)) };
		}
		throw new Error("");
	}

	export const toJson = memoize(function(state: State): JSON {
		switch (state.type) {
			case "map":
				return mapValues(state.children, toJson);
			case "list":
				return state.order.map(key => toJson(state.children[key]));
			default:
				return state.value;
		}
	});

	function get(state: State, path: any): State {
		return toPath(path).reduce(getChild, state);
	}

	export function getChild(state: State, key: string): State {
		ensure(isBranchState(state), "Cannot get child of primitive.");
		ensure(key in state.children);
		return state.children[key];
	}

	export function set(state: State, path: string, value: Primitive): void {
		const target = get(state, path);
		ensure(!isBranchState(target), "Cannot set value of branch.");
		target.value = value;
	}

	const MAX = Number.MAX_SAFE_INTEGER;
	export function add(state: State, path: string, value: State, index: number = MAX): void {
		const pathArray = toPath(path);
		const target = get(state, initial(pathArray));
		ensure(isBranchState(target), "Cannot add property to primitive.");
		const key = last(pathArray);
		ensure(key, "Cannot add property at empty path.");
		target.children[key] = value;
		if (target.type === "list") target.order.splice(index, 0, key);
	}

	export function remove(state: State, path: any): void {
		const pathArray = toPath(path);
		const target = get(state, initial(pathArray));
		ensure(isBranchState(target), "Cannot remove property to primitive.");
		const key = last(pathArray);
		ensure(key, "Cannot remove property at empty path.");
		delete target.children[key];
	}

	export function blur(state: State, path: any, user: string): void {
		const target = get(state, path);
		ensure(!isBranchState(target), "Cannot blur branch.");
		target.touched = true;
		target.focusedBy.filter(u => u !== user);
	}

	export function focus(state: State, path: string, user: string): void {
		const target = get(state, path);
		ensure(!isBranchState(target), "Cannot focus branch.");
		target.focusedBy.push(user);
	}
}
