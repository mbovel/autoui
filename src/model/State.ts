import {
	ensure,
	JSONType,
	JSONObject,
	JSONArray,
	Primitive,
	isPrimitive,
	isDefined
} from "./utils";
import { Action } from "./Action";
import toPath from "lodash/toPath";
import isArray from "lodash/isArray";
import zipObject from "lodash/zipObject";
import mapValues from "lodash/mapValues";
import keys from "lodash/keys";

export interface BaseState {
	focusedBy?: string[];
	selectedBy?: string[];
	touched?: boolean;
}

export interface PrimitiveState<D extends Primitive = Primitive> extends BaseState {
	value: D;
}

export interface ObjectState<D extends JSONObject = JSONObject> extends BaseState {
	children: { [K in keyof D]: StateOf<D[K]> };
}

export interface ArrayState<D extends JSONArray = JSONArray> extends BaseState {
	children: { [key: string]: StateOf<D[number]> };
	order: string[];
}

// See https://github.com/Microsoft/TypeScript/issues/29368#issuecomment-453529532
export type StateOf<D extends JSONType = JSONType> = [D] extends [Primitive]
	? PrimitiveState<D>
	: [D] extends [JSONObject]
	? ObjectState<D>
	: [D] extends [JSONArray]
	? ArrayState<D>
	: State;

export type State = PrimitiveState | ObjectState | ArrayState;

export function isPrimitiveState(state: State): state is PrimitiveState {
	return isDefined((state as any).value);
}

export function isArrayState(state: State): state is ArrayState {
	return (state as any).children && (state as any).order;
}

export function isObjectState(state: State): state is ObjectState {
	return (state as any).children && !(state as any).order;
}

export function isBranchState(state: State): state is ObjectState | ArrayState {
	return (state as any).children;
}

export function mutateState(state: State, action: Action, user: string = "me"): void {
	switch (action.type) {
		case "set":
			set(state, action.path, action.data);
			return;
		case "insertAfter":
			insertAfter(state, action.path, action.key, action.data, action.index);
			return;
		case "remove":
			remove(state, action.path, action.key);
			return;
		case "sort":
			sort(state, action.path, action.compare);
			return;
		case "focus":
			focus(state, action.path, user);
			return;
		case "blur":
			blur(state, action.path, user);
			return;
	}
}

export function stateFromJson<D extends JSONType>(data: D): StateOf<D>;
export function stateFromJson(data: JSONType): State;
export function stateFromJson(data: JSONType): State {
	if (isPrimitive(data)) {
		return { value: data };
	} else if (isArray(data)) {
		const order = keys(data);
		return { order, children: zipObject(order, data.map(stateFromJson)) };
	} else {
		return { children: mapValues(data, stateFromJson) };
	}
}

export function stateToJson<D extends JSONType>(state: StateOf<D>): D;
export function stateToJson(state: State): JSONType;
export function stateToJson(state: State): JSONType {
	if (isPrimitiveState(state)) {
		return state.value;
	} else if (isArrayState(state)) {
		return state.order.map(key => stateToJson<JSONType>(state.children[key]));
	} else {
		return mapValues(state.children, stateToJson);
	}
}

function set(state: State, path: string, value: Primitive): void {
	const target = get(state, path);
	ensure(!isBranchState(target), "Cannot set value of branch.");
	target.value = value;
}

function get(state: State, path: any): State {
	return toPath(path).reduce(getChildState, state);
}

function getChildState(state: State, key: string): State {
	ensure(isBranchState(state), "Cannot get child of primitive.");
	ensure(key in state.children);
	return state.children[key];
}

const MAX = Number.MAX_SAFE_INTEGER;
function insertAfter(
	state: State,
	path: string,
	key: string,
	value: State,
	index: number = MAX
): void {
	const target = get(state, path);
	ensure(isArrayState(target), "Can only insertAfter in array.");
	target.children[key] = value;
	target.order.splice(index, 0, key);
}

function remove(state: State, path: any, key: string): void {
	const target = get(state, path);
	ensure(isArrayState(target), "Can only remove in array.");
	ensure(key, "Cannot remove property at empty path.");
	delete target.children[key];
	target.order.splice(target.order.indexOf(key), 1);
}

export function sort(state: State, path: any, compare: (a: any, b: any) => number): void {
	const target = get(state, path);
	ensure(isArrayState(target), "Cannot only sort array.");
	target.order.sort((aKey, bKey) => compare(target.children[aKey], target.children[bKey]));
}

export function touch(state: State, path: string) {
	let current = state;
	for (const key of toPath(path)) {
		if (!current.touched) current.touched = true;
		current = getChildState(current, key);
	}
	if (!current.touched) current.touched = true;
}

export function blur(state: State, path: any, user: string): void {
	const target = get(state, path);
	touch(state, path);
	target.focusedBy = target.focusedBy?.filter(u => u !== user);
}

export function focus(state: State, path: string, user: string): void {
	const target = get(state, path);
	if (!target.focusedBy) target.focusedBy = [];
	target.focusedBy.push(user);
}

/*
function unselect(state: State, path: any, user: string): void {
	const target = get(state, path);
	target.selectedBy = target.selectedBy?.filter(u => u !== user);
}

function select(state: State, path: any, user: string): void {
	const target = get(state, path);
	if (!target.selectedBy) target.selectedBy = [];
	target.selectedBy.push(user);
}
*/
