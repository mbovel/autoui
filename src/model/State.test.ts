import { fromJson, touch } from "./State";

describe("fromJson", () => {
	test("correctly converts string", () => {
		expect(fromJson("test")).toMatchSnapshot();
	});

	test("correctly converts number", () => {
		expect(fromJson(2)).toMatchSnapshot();
	});

	test("correctly converts array", () => {
		expect(fromJson([1, 2, 3])).toMatchSnapshot();
	});

	test("correctly converts object", () => {
		expect(fromJson({ prop1: 1, prop2: 2 })).toMatchSnapshot();
	});
});

describe("touch", () => {
	test("works on single string", () => {
		const state = fromJson(2);
		touch(state, "");
		expect(state.touched).toBe(true);
	});

	test("touch all ancestors", () => {
		const state = fromJson({ foo: 2 });
		touch(state, "foo");
		expect(state.touched).toBe(true);
		expect(state.children.foo.touched).toBe(true);
	});

	test("doesn't touch descendants", () => {
		const state = fromJson({ foo: 2 });
		touch(state, "");
		expect(state.touched).toBe(true);
		expect(state.children.foo.touched).toBeFalsy();
	});
});
