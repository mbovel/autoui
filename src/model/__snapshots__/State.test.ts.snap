// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromJson correctly converts array 1`] = `
Object {
  "children": Object {
    "0": Object {
      "value": 1,
    },
    "1": Object {
      "value": 2,
    },
    "2": Object {
      "value": 3,
    },
  },
  "order": Array [
    "0",
    "1",
    "2",
  ],
}
`;

exports[`fromJson correctly converts number 1`] = `
Object {
  "value": 2,
}
`;

exports[`fromJson correctly converts object 1`] = `
Object {
  "children": Object {
    "prop1": Object {
      "value": 1,
    },
    "prop2": Object {
      "value": 2,
    },
  },
}
`;

exports[`fromJson correctly converts string 1`] = `
Object {
  "value": "test",
}
`;
